src/components/timeline/TimelineMarkerCanvas.js

```diff
@@ -155,37 +204,19 @@ class TimelineMarkerCanvas extends PureComponent {
 
           const tracingMarkerIndex = markerTiming.index[i];
           const isHovered = hoveredItem === tracingMarkerIndex;
-          ctx.fillStyle = isHovered ? 'Highlight' : '#8296cb';
-
-          if (w >= h) {
-            this.drawRoundedRect(ctx, x, y + 1, w, h - 1, 1);
-
-            const text = markerTiming.label[i];
-            // Draw the text label
-            // TODO - L10N RTL.
-            // Constrain the x coordinate to the leftmost area.
-            const x2: CssPixels = Math.max(x, 0) + TEXT_OFFSET_START;
-            const w2: CssPixels = Math.max(0, w - (x2 - x));
-
-            if (w2 > textMeasurement.minWidth) {
-              const fittedText = textMeasurement.getFittedText(text, w2);
-              if (fittedText) {
-                ctx.fillStyle = isHovered ? 'HighlightText' : '#ffffff';
-                ctx.fillText(fittedText, x2, y + TEXT_OFFSET_TOP);
-              }
-            }
+          const text = markerTiming.label[i];
+          if (isHovered) {
+            hoveredElement = { x, y, w, h, text };
           } else {
-            ctx.beginPath();
-            ctx.arc(
-              x + w / 2, // x
-              y + h / 2, // y
-              h * MARKER_DOT_RADIUS, // radius
-              0, // arc start
-              TWO_PI // arc end
-            );
-            ctx.fill();
+            this.drawOneMarker(ctx, { x, y, w, h, text });
           }
         }
+        if (hoveredElement) {
+          this.drawOneMarker(ctx, hoveredElement, {
```

Comment:
I'm concerned about allocating objects here. I think we'll hit some GC pause issues. Would you mind changing this into not creating new object allocations?

Reply:
oh yeah, I think you're right. I actually hit a lot of GC pause issues while working on this and I never considered it could be because of this.

I think it's OK doing it only for the hovered element because it's at most once per drawing; but for the first call it could be massive.

Reply:
There was actually 2 allocated objects: the drawing information _and_ the color information -- that last one was hidden as a default parameter, which in JavaScript is recreated for each call ;)